vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace rec Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Windows.Controls
+open System.Windows.Media
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text.Formatting
+open System.ComponentModel.Composition
+open Microsoft.VisualStudio.Utilities
+open Microsoft.CodeAnalysis
+
+type internal CodeLensAdornment(view:IWpfTextView) as this=
```

Comment:
Yeah, you folks are building essentially F#-only CodeLens-like feature from scratch, not using editor's CodeLens API and implementation.
That might be OK if that's what you want (to workaround CodeLens being Pro+ feature for example).
But it's definitely going to be more expensive to implement and definitely will be inconsistent with editor's CodeLens user experience wise.
We are currently considering making CodeLens API public (which technically is already public, but just defined in an internal assembly) and the only limiting factor is the lack of time (we are heads down with significant VS update work) so I'm pretty confident it will be public eventually. If you'd like to go that route I can show you how it can be done.